### PR TITLE
Ds/gmsh support

### PR DESCRIPTION
**Pull request by dsteinmo**

### CODE REVIEW

File: include/MeshManager.hpp [hunk 1]
```diff
@@ -29,6 +30,11 @@ namespace blitzdg {
       index_vec_smart_ptr EToV;
       index_vec_smart_ptr ElementPartitionMap;
       index_vec_smart_ptr VertexPartitionMap;
+
+
+      // Helper method to convert vector of strings to vector of ints
+      // representing Gmsh element data.
+      std::vector<index_type> parseElem(std::vector<std::string> input);
```
Comment (k7miller) [inline include/MeshManager.hpp]: Pass the input by const reference `const std::vector<std::string>& input`.

File: src/MeshManager.cpp [hunk 2]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+        std::vector<index_type> elem(input.size());
+        index_type count = 0;
+        for (auto &s : input) {
+            std::stringstream parser(s);
+            index_type x = 0;
+            parser >> x;
+            elem[count] = x;
+            ++count;
+        }
+        return std::move(elem);
```
Comment (k7miller) [inline src/MeshManager.cpp]: You don't need std::move here.

File: src/MeshManager.cpp [hunk 3]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
         ElementPartitionMap{ nullptr }, VertexPartitionMap{ nullptr }
     {}
 
+
+    std::vector<index_type> MeshManager::parseElem(std::vector<string> input) {
+
+        std::vector<index_type> elem(input.size());
+        index_type count = 0;
+        for (auto &s : input) {
```
Comment (k7miller) [inline src/MeshManager.cpp]: I would use `push_back` to build elems here. Something like
```
std::vector<index_type> elems;
elems.reserve(input.size());
for (const auto& s : input)
    elems.push_back(std::stoi(s));
```

File: src/MeshManager.cpp [hunk 4]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+    }
+
     void MeshManager::readMesh(const string& gmshInputFile) {
-        throw runtime_error("MeshManager::readMesh not implemented");
+        // gmsh .msh files are csv's with special headers/separators that begin
+        // with either "'s or $'s.
+
+        CSVFileReader csvReader(gmshInputFile);
+
+        string line="";
```
Comment (k7miller) [inline src/MeshManager.cpp]: Just use `string line`.

File: src/MeshManager.cpp [hunk 5]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+        csvReader.readLine(line);
+
+        // parse headers
+        if (line.compare("$MeshFormat") != 0)
```
Comment (k7miller) [inline src/MeshManager.cpp]: Use `line != "$MeshFormat"` instead of compare.

File: src/MeshManager.cpp [hunk 6]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+            throw runtime_error("Only ASCII-type Gmsh formats are supported.");
+
+        if(floatSize != 8)
+            throw runtime_error("Only 8-byte reals in Gmsh files are supported!");
+
+        // This should skip the '$EndMeshFormat' line.
+        csvReader.skipLines(1);
+
+        csvReader.readLine(line);
+        if (line.compare("$Nodes") != 0)
```
Comment (k7miller) [inline src/MeshManager.cpp]: Same as above, use `line != "$Nodes"`.

File: src/MeshManager.cpp [hunk 7]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+            Vref((nodeNum-1)*3)   = vx;
+            Vref((nodeNum-1)*3+1) = vy;
+            Vref((nodeNum-1)*3+2) = vz;
+        }
+
+        // This should skip the '$EndNodes' line.
+        csvReader.skipLines(1);
+        
+        csvReader.readLine(line);
+        if (line.compare("$Elements") != 0)
```
Comment (k7miller) [inline src/MeshManager.cpp]: Same as above use `line != "$Elements"`.

File: src/MeshManager.cpp [hunk 8]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+                    throw runtime_error("Incorrect Element Type for point element!");
+                points.push_back(stoi(elementInfo[3]));
+            }
+
+            if (numLocalVerts == 2) {
+                if (elemType !=1)
+                    throw runtime_error("Incorrect Element Type for line element!");
+                
+                std::vector<index_type> lineElem = parseElem(elementInfo);
+                lines.push_back(lineElem);
```
Comment (k7miller) [inline src/MeshManager.cpp]: Replace these two lines with 'lines.push_back(parseElem(elementInfo));`.

File: src/MeshManager.cpp [hunk 9]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+            }
+
+            if (numLocalVerts == 3) {
+                if (elemType != 2)
+                    throw runtime_error("Incorrect Element Type for triangle element!");
+
+                std::vector<index_type> triElem = parseElem(elementInfo);
+                tris.push_back(triElem);
```
Comment (k7miller) [inline src/MeshManager.cpp]: Replace these lines with `tris.push_back(parseElem(elementInfo));`

File: src/MeshManager.cpp [hunk 10]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+            }
+
+            if (numLocalVerts == 4) {
+                if (elemType != 3)
+                    throw runtime_error("Incorrect Element Type for quadrangle element!");
+                
+                std::vector<index_type> quadElem = parseElem(elementInfo);
+                quads.push_back(quadElem);
```
Comment (k7miller) [inline src/MeshManager.cpp]: Replace these lines with `quads.push_back(parseElem(elementInfo));`

File: src/MeshManager.cpp [hunk 11]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+        std::vector<index_type> points;
+        std::vector<std::vector<index_type>> lines;
+        std::vector<std::vector<index_type>> tris;
+        std::vector<std::vector<index_type>> quads;
+
+        for(index_type i =0; i < numElementRows; ++i) {
+            csvReader.readLine(line);
+            csvReader.tokenizeLine(line, elementInfo);
+
+            index_type numCols = elementInfo.size();
```
Comment (k7miller) [inline src/MeshManager.cpp]: Consider adding a cast to `index_type` here.

File: src/MeshManager.cpp [hunk 12]
```diff
@@ -47,8 +48,164 @@ namespace blitzdg {
...
+                std::vector<index_type> quadElem = parseElem(elementInfo);
+                quads.push_back(quadElem);
+            }
+        }
+
+        if (quads.size() > 0)
+            throw runtime_error("Quadrangle elements currently not supported by blitzdg!");
+
+        // Allocate storage EToV.
+        index_type K = tris.size();
```
Comment (k7miller) [inline src/MeshManager.cpp]: Consider adding a cast to `index_type` here.

**Review by k7miller — APPROVED** [review]
Looks good buddy. I left some comments for you to consider.

File: src/MeshManager.cpp (hunk 2 above)
Comment (dsteinmo) [reply src/MeshManager.cpp]: Thanks

**Comment by dsteinmo** [comment]
:shipit:

**Pull request merged**